packages/flutter/lib/src/painting/image_stream.dart

```diff
@@ -196,54 +217,116 @@ abstract class ImageStreamCompleter extends Diagnosticable {
   /// occurred. If the listener is added within a render object paint function,
   /// then use this flag to avoid calling [RenderObject.markNeedsPaint] during
   /// a paint.
-  void addListener(ImageListener listener) {
-    _listeners.add(listener);
-    if (_current != null) {
+  void addListener(ImageListener listener, { ImageErrorListener onError }) {
+    _listeners[listener] = onError;
+    if (_currentImage != null) {
       try {
-        listener(_current, true);
+        listener(_currentImage, true);
       } catch (exception, stack) {
-        _handleImageError('by a synchronously-called image listener', exception, stack);
+        reportError(
+          context: 'by a synchronously-called image listener',
+          exception: exception,
+          stack: stack,
+        );
       }
     }
+    if (_currentError != null && onError != null) {
+      try {
+          onError(_currentError.exception, _currentError.stack);
+        } catch (exception, stack) {
+          reportError(
+            context: 'by a synchronously-called image error listener',
+            exception: exception,
+            stack: stack,
+            // Error listeners themselves failed. Don't feed back to listeners.
+            skipListeners: true,
+          );
+        }
+    }
   }
 
-  /// Stop listening for new concrete [ImageInfo] objects.
+  /// Stop listening for new concrete [ImageInfo] objects and errors from
+  /// its associated [ImageErrorListener].
   void removeListener(ImageListener listener) {
     _listeners.remove(listener);
   }
 
   /// Calls all the registered listeners to notify them of a new image.
   @protected
   void setImage(ImageInfo image) {
-    _current = image;
+    _currentImage = image;
```

Comment:
Should this clear an error in `_currentError` ? Since a listener could be called multiple times, if a prior completer has an error wouldn't all subsequent images complete with an error?

Reply:
Ya, I wasn't sure if I covered all the scenarios either. Though I was thinking, if it's a multiframe image, and one of the frames failed but subsequent frames succeeded, we'd still want to make it visible.

cc @amirh for thoughts

Perhaps we need another API on the ImageStreamCompleter to clear itself when passing into setCompleter of an ImageStream?

Reply:
Turned out this will never happen. A failed decode will never try to decode the next frame again.